### DISCUSSION TITLE
외부 domain의 repository와 service 사용

### DISCUSSION THREAD

**Discussion by Kimhan-nah** [GhostPangPang/GhostPong#127] [General]
### 문제
- #48 
- 위의 discussion에서 임시로 repository 접근하는 걸로 결정하고 구현했지만, 현재 service, repository 모두 사용하고 있음
- 한가지 방법으로 정할 필요가 있을 것 같다!

### 제안

- GTP 선생님에게 여러 차례 질문을 해본 결과 대부분의 답변이 `repository`보다 `service`로 사용하는 것이 구조 단순화와 유지 보수를 위해 바람직하다고 합니다.
<img width="672" alt="스크린샷 2023-04-24 오후 2 39 08" src="https://user-images.githubusercontent.com/33301153/233908789-b1735a6a-e6da-489c-a0b4-1b38d6cc41d2.png">

- _하지만 `service`를 사용하게 되면 순환 참조 문제가 발생하게 되는데 이것은 구조의 문제일 수도 있... (해결 방법이 있다고 함ㅎㅎ)_
- error handling의 경우 함수명을 `IfNotExsit` 혹은 `IfExist`와 같이 명시해서 각각 구현하는 것도 좋을 것 같아용

다양한 의견 받습니다~!

**Comment by anso33** [comment]
이거 제가 위키에 제목만 적고 task 미뤄놓은 주제인데 시험 빨리 끝내면 열심히 찾아보고 제안하겠습니다. 그렇지만 궁금하셔서 먼저 찾아보고 제안하겠다!! 대환영입니다....

**Reply by anso33** [reply]
현재 service를 사용하는 경우는 그 메서드를 해당 모듈에서도 사용하고 외부 모듈에서도 사용할 때
repository를 사용할 때는 그 메서드를 외부 모듈에서만 사용할 때입니다. (아마두)

통일할 필요는 있을 것 같아서 고민해보고, 찾아보고 리팩토링 제안하겠숩니당

**Reply by anso33** [reply]
근데 B에서 A서비스를 이미 가져다 쓰고 있는데 A에서는 사용하지 않을 메서드다...! 이런 경우  A서비스도 가져오고 Arepository도 가져오기 이상해서 AService에 구현하고 A Service만 가져다 쓰는 방식으로 작성했(던것 같)습니다.

**Comment by jis-kim** [comment]
커스텀 레포지토리로 또 마음이 끌립니다... 데이터를 분리해버리고 싶은 이 맘 뭘까?

**Reply by jis-kim** [reply]
만약 커스텀 레포지토리를 사용할 경우, 트랜잭션 관리에 문제가 생기므로 entity 단위로 나눌 수 없습니다. ex) 유저 생성 시 user, auth 모두에 접근해야 함
서비스 레이어 안에서 메소드를 나눌 경우 문제점도 현재 typeorm 에서 관리하는 레포지토리 별로 도메인이 칼같이 나뉠수 없어서 생기는 일 같아요.
